From 8246858d6ef8a98aed2034aaf015e1ee69d9b65e Mon Sep 17 00:00:00 2001
From: Arun Raghavan <arunsr@codeaurora.org>
Date: Tue, 29 May 2018 09:05:03 +0530
Subject: [PATCH 3/3] audioringbuffer: Add support for IEC61937-payloaded data

Represents unknown content in a IEC61937 payload. Underlying data rate
might need to be signalled via the ringbuffer caps.
---
 gst-libs/gst/audio/gstaudioringbuffer.c | 19 ++++++++++++++++++-
 gst-libs/gst/audio/gstaudioringbuffer.h |  4 +++-
 2 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/gst-libs/gst/audio/gstaudioringbuffer.c b/gst-libs/gst/audio/gstaudioringbuffer.c
index d3c13152f..40de855aa 100644
--- a/gst-libs/gst/audio/gstaudioringbuffer.c
+++ b/gst-libs/gst/audio/gstaudioringbuffer.c
@@ -333,7 +333,24 @@ gst_audio_ring_buffer_parse_caps (GstAudioRingBufferSpec * spec, GstCaps * caps)
 
     gst_structure_get_int (structure, "channels", &info.channels);
     spec->type = GST_AUDIO_RING_BUFFER_FORMAT_TYPE_WMA;
-    info.bpf = 1;
+  } else if (g_str_equal (mimetype, "audio/x-iec61937")) {
+    const gchar *payload_type;
+
+    if (!(gst_structure_get_int (structure, "rate", &info.rate)))
+      goto parse_error;
+
+    payload_type = gst_structure_get_string (structure, "payload-type");
+
+    if (!payload_type || g_strcmp0 (payload_type, "normal") == 0)
+      info.bpf = 4;             /* stereo s16 equivalent */
+    else if (g_strcmp0 (payload_type, "4x") == 0)
+      info.bpf = 16;            /* 4 * stereo s16 equivalent */
+    else if (g_strcmp0 (payload_type, "hbr") == 0)
+      info.bpf = 16;            /* 8ch s16 equivalent, rate accounts for higher bitrate */
+    else
+      goto parse_error;
+
+    spec->type = GST_AUDIO_RING_BUFFER_FORMAT_TYPE_IEC61937;
   } else {
     goto parse_error;
   }
diff --git a/gst-libs/gst/audio/gstaudioringbuffer.h b/gst-libs/gst/audio/gstaudioringbuffer.h
index 3194e0fb8..9d9991fe7 100644
--- a/gst-libs/gst/audio/gstaudioringbuffer.h
+++ b/gst-libs/gst/audio/gstaudioringbuffer.h
@@ -91,6 +91,7 @@ typedef enum {
  * @GST_AUDIO_RING_BUFFER_FORMAT_TYPE_VORBIS: samples in Vorbis format
  * @GST_AUDIO_RING_BUFFER_FORMAT_TYPE_ALAC: samples in ALAC format
  * @GST_AUDIO_RING_BUFFER_FORMAT_TYPE_WMA: samples in WMA format
+ * @GST_AUDIO_RING_BUFFER_FORMAT_TYPE_IEC61937: samples in IEC61937 frames (Since 1.16)
  *
  * The format of the samples in the ringbuffer.
  */
@@ -113,7 +114,8 @@ typedef enum
   GST_AUDIO_RING_BUFFER_FORMAT_TYPE_FLAC,
   GST_AUDIO_RING_BUFFER_FORMAT_TYPE_VORBIS,
   GST_AUDIO_RING_BUFFER_FORMAT_TYPE_ALAC,
-  GST_AUDIO_RING_BUFFER_FORMAT_TYPE_WMA
+  GST_AUDIO_RING_BUFFER_FORMAT_TYPE_WMA,
+  GST_AUDIO_RING_BUFFER_FORMAT_TYPE_IEC61937,
 } GstAudioRingBufferFormatType;
 
 /**
-- 
2.19.1

